File: rust/src/protocol.rs
//! Raw iLink Bot API HTTP calls.

use base64::Engine;
use rand::RngCore;
use reqwest::Client;
use serde::Deserialize;
use serde_json::{json, Value};
use std::time::Duration;
use uuid::Uuid;

use crate::error::{Result, WeChatBotError};
use crate::types::*;

pub const DEFAULT_BASE_URL: &str = "https://ilinkai.weixin.qq.com";
pub const CDN_BASE_URL: &str = "https://novac2c.cdn.weixin.qq.com/c2c";
pub const CHANNEL_VERSION: &str = "2.0.0";

/// Generate the X-WECHAT-UIN header value.
pub fn random_wechat_uin() -> String {
    let mut buf = [0u8; 4];
    rand::thread_rng().fill_bytes(&mut buf);
    let val = u32::from_be_bytes(buf);
    base64::engine::general_purpose::STANDARD.encode(val.to_string())
}

/// QR code response.
#[derive(Debug, Deserialize)]
pub struct QrCodeResponse {
    pub qrcode: String,
    pub qrcode_img_content: String,
}

/// QR status response.
#[derive(Debug, Deserialize)]
pub struct QrStatusResponse {
    pub status: String,
    pub bot_token: Option<String>,
    pub ilink_bot_id: Option<String>,
    pub ilink_user_id: Option<String>,
    pub baseurl: Option<String>,
}

/// Get updates response.
#[derive(Debug, Deserialize)]
pub struct GetUpdatesResponse {
    #[serde(default)]
    pub ret: Option<i32>,
    #[serde(default)]
    pub msgs: Vec<WireMessage>,
    #[serde(default)]
    pub get_updates_buf: String,
    pub errcode: Option<i32>,
    pub errmsg: Option<String>,
}

/// Get config response.
#[derive(Debug, Deserialize)]
pub struct GetConfigResponse {
    pub typing_ticket: Option<String>,
}

/// Low-level iLink API client.
pub struct ILinkClient {
    http: Client,
}

impl ILinkClient {
    pub fn new() -> Self {
        Self {
            http: Client::builder()
                .timeout(Duration::from_secs(45))
                .build()
                .unwrap(),
        }
    }

    pub async fn get_qr_code(&self, base_url: &str) -> Result<QrCodeResponse> {
        let url = format!("{}/ilink/bot/get_bot_qrcode?bot_type=3", base_url);
        let resp = self.http.get(&url).send().await?;
        Ok(resp.json().await?)
    }

    pub async fn poll_qr_status(&self, base_url: &str, qrcode: &str) -> Result<QrStatusResponse> {
        let url = format!(
            "{}/ilink/bot/get_qrcode_status?qrcode={}",
            base_url,
            urlencoding::encode(qrcode)
        );
        let resp = self
            .http
            .get(&url)
            .header("iLink-App-ClientVersion", "1")
            .send()
            .await?;
        Ok(resp.json().await?)
    }

    pub async fn get_updates(
        &self,
        base_url: &str,
        token: &str,
        cursor: &str,
    ) -> Result<GetUpdatesResponse> {
        let body = json!({
            "get_updates_buf": cursor,
            "base_info": { "channel_version": CHANNEL_VERSION }
        });
        let resp = self.api_post(base_url, "/ilink/bot/getupdates", token, &body, 45).await?;
        let result: GetUpdatesResponse = serde_json::from_value(resp)?;
        if result.ret.is_some_and(|ret| ret != 0) {
            let ret = result.ret.unwrap_or_default();
            let code = result.errcode.unwrap_or(ret);
            let msg = result.errmsg.unwrap_or_else(|| format!("ret={}", ret));
            return Err(WeChatBotError::Api {
                message: msg,
                http_status: 200,
                errcode: code,
            });
        }
        Ok(result)
    }

    pub async fn send_message(&self, base_url: &str, token: &str, msg: &Value) -> Result<()> {
        let body = json!({
            "msg": msg,
            "base_info": { "channel_version": CHANNEL_VERSION }
        });
        self.api_post(base_url, "/ilink/bot/sendmessage", token, &body, 15).await?;
        Ok(())
    }

    pub async fn get_config(
        &self,
        base_url: &str,
        token: &str,
        user_id: &str,
        context_token: &str,
    ) -> Result<GetConfigResponse> {
        let body = json!({
            "ilink_user_id": user_id,
            "context_token": context_token,
            "base_info": { "channel_version": CHANNEL_VERSION }
        });
        let resp = self.api_post(base_url, "/ilink/bot/getconfig", token, &body, 15).await?;
        Ok(serde_json::from_value(resp)?)
    }

    pub async fn send_typing(
        &self,
        base_url: &str,
        token: &str,
        user_id: &str,
        ticket: &str,
        status: i32,
    ) -> Result<()> {
        let body = json!({
            "ilink_user_id": user_id,
            "typing_ticket": ticket,
            "status": status,
            "base_info": { "channel_version": CHANNEL_VERSION }
        });
        self.api_post(base_url, "/ilink/bot/sendtyping", token, &body, 15).await?;
        Ok(())
    }

    async fn api_post(
        &self,
        base_url: &str,
        endpoint: &str,
        token: &str,
        body: &Value,
        timeout_secs: u64,
    ) -> Result<Value> {
        let url = format!("{}{}", base_url, endpoint);
        let resp = self
            .http
            .post(&url)
            .timeout(Duration::from_secs(timeout_secs))
            .header("Content-Type", "application/json")
            .header("AuthorizationType", "ilink_bot_token")
            .header("Authorization", format!("Bearer {}", token))
            .header("X-WECHAT-UIN", random_wechat_uin())
            .json(body)
            .send()
            .await?;

        let status = resp.status().as_u16();
        let text = resp.text().await?;
        let value: Value = serde_json::from_str(&text).unwrap_or(json!({}));

        if status >= 400 {
            return Err(WeChatBotError::Api {
                message: value["errmsg"].as_str().unwrap_or(&text).to_string(),
                http_status: status,
                errcode: value["errcode"].as_i64().unwrap_or(0) as i32,
            });
        }

        Ok(value)
    }

    /// Request an upload URL for CDN media upload.
    pub async fn get_upload_url(
        &self,
        base_url: &str,
        token: &str,
        params: &Value,
    ) -> Result<GetUploadUrlResponse> {
        let mut body = params.clone();
        body["base_info"] = json!({ "channel_version": CHANNEL_VERSION });
        let resp = self.api_post(base_url, "/ilink/bot/getuploadurl", token, &body, 15).await?;
        Ok(serde_json::from_value(resp)?)
    }
}

/// Get upload URL response.
#[derive(Debug, Deserialize)]
pub struct GetUploadUrlResponse {
    pub upload_param: Option<String>,
}

/// Build a media message payload.
pub fn build_media_message(user_id: &str, context_token: &str, item_list: Vec<Value>) -> Value {
    json!({
        "from_user_id": "",
        "to_user_id": user_id,
        "client_id": Uuid::new_v4().to_string(),
        "message_type": 2,
        "message_state": 2,
        "context_token": context_token,
        "item_list": item_list
    })
}

/// Build a text message payload.
pub fn build_text_message(user_id: &str, context_token: &str, text: &str) -> Value {
    json!({
        "from_user_id": "",
        "to_user_id": user_id,
        "client_id": Uuid::new_v4().to_string(),
        "message_type": 2,
        "message_state": 2,
        "context_token": context_token,
        "item_list": [{ "type": 1, "text_item": { "text": text } }]
    })
}

File: reference-sdks/python/wechatbot/protocol.py
"""Raw iLink Bot API HTTP calls."""

from __future__ import annotations

import base64
import json
import os
import struct
from typing import Any
from urllib.parse import quote
from uuid import uuid4

import aiohttp

from .errors import ApiError
from .types import MediaType, MessageItemType, MessageState, MessageType

DEFAULT_BASE_URL = "https://ilinkai.weixin.qq.com"
CDN_BASE_URL = "https://novac2c.cdn.weixin.qq.com/c2c"
CHANNEL_VERSION = "2.0.0"


def random_wechat_uin() -> str:
    val = struct.unpack(">I", os.urandom(4))[0]
    return base64.b64encode(str(val).encode("utf-8")).decode("ascii")


def auth_headers(token: str) -> dict[str, str]:
    return {
        "Content-Type": "application/json",
        "AuthorizationType": "ilink_bot_token",
        "Authorization": f"Bearer {token}",
        "X-WECHAT-UIN": random_wechat_uin(),
    }


def _base_info() -> dict[str, str]:
    return {"channel_version": CHANNEL_VERSION}


async def _parse_response(resp: aiohttp.ClientResponse, label: str) -> dict[str, Any]:
    text = await resp.text()
    payload: dict[str, Any] = json.loads(text) if text else {}

    if resp.status >= 400:
        msg = payload.get("errmsg") or f"{label} failed with HTTP {resp.status}"
        raise ApiError(
            msg,
            http_status=resp.status,
            errcode=payload.get("errcode", 0),
            payload=payload,
        )

    ret = payload.get("ret")
    if isinstance(ret, int) and ret != 0:
        code = payload.get("errcode", ret)
        msg = payload.get("errmsg") or f"{label} failed (ret={ret})"
        raise ApiError(msg, http_status=resp.status, errcode=code, payload=payload)

    return payload


class ILinkApi:
    """Low-level iLink API client. Each method maps 1:1 to an endpoint."""

    def __init__(self) -> None:
        self._timeout = aiohttp.ClientTimeout(total=45)

    async def get_qr_code(self, base_url: str) -> dict[str, Any]:
        url = f"{base_url}/ilink/bot/get_bot_qrcode?bot_type=3"
        async with aiohttp.ClientSession() as session:
            async with session.get(url) as resp:
                return await _parse_response(resp, "get_bot_qrcode")

    async def poll_qr_status(self, base_url: str, qrcode: str) -> dict[str, Any]:
        url = f"{base_url}/ilink/bot/get_qrcode_status?qrcode={quote(qrcode, safe='')}"
        async with aiohttp.ClientSession() as session:
            async with session.get(
                url, headers={"iLink-App-ClientVersion": "1"}
            ) as resp:
                return await _parse_response(resp, "get_qrcode_status")

    async def get_updates(
        self, base_url: str, token: str, cursor: str
    ) -> dict[str, Any]:
        body = {"get_updates_buf": cursor, "base_info": _base_info()}
        return await self._post(base_url, "/ilink/bot/getupdates", token, body, 45)

    async def send_message(
        self, base_url: str, token: str, msg: dict[str, Any]
    ) -> dict[str, Any]:
        body = {"msg": msg, "base_info": _base_info()}
        return await self._post(base_url, "/ilink/bot/sendmessage", token, body)

    async def get_config(
        self, base_url: str, token: str, user_id: str, context_token: str
    ) -> dict[str, Any]:
        body = {
            "ilink_user_id": user_id,
            "context_token": context_token,
            "base_info": _base_info(),
        }
        return await self._post(base_url, "/ilink/bot/getconfig", token, body)

    async def send_typing(
        self,
        base_url: str,
        token: str,
        user_id: str,
        ticket: str,
        status: int,
    ) -> dict[str, Any]:
        body = {
            "ilink_user_id": user_id,
            "typing_ticket": ticket,
            "status": status,
            "base_info": _base_info(),
        }
        return await self._post(base_url, "/ilink/bot/sendtyping", token, body)

    async def get_upload_url(
        self,
        base_url: str,
        token: str,
        params: dict[str, Any],
    ) -> dict[str, Any]:
        body = {**params, "base_info": _base_info()}
        return await self._post(base_url, "/ilink/bot/getuploadurl", token, body)

    @staticmethod
    def build_media_message(
        user_id: str,
        context_token: str,
        item_list: list[dict[str, Any]],
    ) -> dict[str, Any]:
        return {
            "from_user_id": "",
            "to_user_id": user_id,
            "client_id": str(uuid4()),
            "message_type": MessageType.BOT,
            "message_state": MessageState.FINISH,
            "context_token": context_token,
            "item_list": item_list,
        }

    async def _post(
        self,
        base_url: str,
        endpoint: str,
        token: str,
        body: dict[str, Any],
        timeout_secs: int = 15,
    ) -> dict[str, Any]:
        url = f"{base_url.rstrip('/')}{endpoint}"
        timeout = aiohttp.ClientTimeout(total=timeout_secs)
        async with aiohttp.ClientSession(timeout=timeout) as session:
            async with session.post(
                url, headers=auth_headers(token), json=body
            ) as resp:
                return await _parse_response(resp, endpoint)

    @staticmethod
    def build_text_message(
        user_id: str, context_token: str, text: str
    ) -> dict[str, Any]:
        return {
            "from_user_id": "",
            "to_user_id": user_id,
            "client_id": str(uuid4()),
            "message_type": MessageType.BOT,
            "message_state": MessageState.FINISH,
            "context_token": context_token,
            "item_list": [
                {"type": MessageItemType.TEXT, "text_item": {"text": text}}
            ],
        }
